### PR TITLE
upload static gpt demo

### PR DESCRIPTION
**Pull request by youth123**
upload static gpt demo

### CODE REVIEW

File: examples/static/gpt/run_static_mp.sh [hunk 1]
```diff
@@ -0,0 +1,44 @@
+set -x
```
Comment (sandyhouse) [inline examples/static/gpt/run_static_mp.sh]: 把run_static_dp_mp.sh run_static_dp_mp.sh run_static_dp.sh整合成run_static.sh一个文件吧，通过参数控制运行哪种模式。

File: examples/static/paddlenlp/transformers/.ipynb_checkpoints/__init__-checkpoint.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (sandyhouse) [inline examples/static/paddlenlp/transformers/.ipynb_checkpoints/__init__-checkpoint.py]: What's this file for?

File: examples/static/paddlenlp/transformers/.ipynb_checkpoints/attention_utils-checkpoint.py [hunk 3]
```diff
@@ -0,0 +1,742 @@
+#   Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (sandyhouse) [inline examples/static/paddlenlp/transformers/.ipynb_checkpoints/attention_utils-checkpoint.py]: What's this file for?

File: examples/static/paddlenlp/transformers/gpt/.ipynb_checkpoints/modeling-checkpoint.py [hunk 4]
```diff
@@ -0,0 +1,911 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (sandyhouse) [inline examples/static/paddlenlp/transformers/gpt/.ipynb_checkpoints/modeling-checkpoint.py]: What's this file for?

File: examples/static/paddlenlp/transformers/gpt/modeling_auto.py [hunk 5]
```diff
@@ -0,0 +1,871 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (sandyhouse) [inline examples/static/paddlenlp/transformers/gpt/modeling_auto.py]: Is this file necessary?

File: examples/static/paddlenlp/transformers/gpt/modeling_serial.py [hunk 6]
```diff
@@ -0,0 +1,799 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (sandyhouse) [inline examples/static/paddlenlp/transformers/gpt/modeling_serial.py]: 整合modeling_manual.py和modeling_serial.py为一个文件？

**Review by sandyhouse — CHANGES_REQUESTED** [review]

File: examples/static/paddlenlp/transformers/.ipynb_checkpoints/__init__-checkpoint.py (hunk 2 above)
Comment (youth123) [reply examples/static/paddlenlp/transformers/.ipynb_checkpoints/__init__-checkpoint.py]: Dummy files.

Comment (youth123) [reply examples/static/paddlenlp/transformers/.ipynb_checkpoints/__init__-checkpoint.py]: I forget to add gitignore file.

File: examples/static/gpt/README.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
+# GPT
+
+## 模型介绍
+GPT-[2](https://cdn.openai.com/better-language-models/language_models_are_unsupervised_multitask_learners.pdf)/[3](https://arxiv.org/pdf/2005.14165.pdf) 是以[Transformer](https://arxiv.org/abs/1706.03762) 解码器为网络基本组件，使用自回归的方式在大规模无标注文本语料上进行预训练得到的语言生成模型。
+
+本项目是语言模型 GPT 的 PaddlePaddle 实现，FOR CE。
```
Comment (sandyhouse) [inline examples/static/gpt/README.md:6]: 去掉For CE吧。我们这个是参考代码，兼备支持CE测试，但不仅仅是为了CE。

File: examples/static/gpt/README.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+```
+mkdir data
+mv train.data.json_ids.npz data
+```
+
+## 飞桨4D混合并行训练
+飞桨4D混合并行，使用sharding、模型并行、流水线并行和数据并行策略，使得训练千亿参数规模的模型成为可能。在本示例中，我们提供了基于飞桨最新混合并行策略的GPT预训练模型。运行下面脚本，指定参数(mp|dp|pp|dp_mp|pp_dp|pp_dp_mp|pp_mp)，即可进行模型预训练对应模型：
+```shell
+sh run_static.sh mp "0,1"
```
Comment (sandyhouse) [inline examples/static/gpt/README.md]: 仅仅指定mp、pp等策略的话，用户怎么指定具体的并行度呢? 比如2卡mp，还是4卡mp。

File: examples/static/gpt/args.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--model_type",
+        default=None,
+        type=str,
+        required=True,
+        help="Model type selected in the list: " )
```
Comment (sandyhouse) [inline examples/static/gpt/args.py:34]: list为空？可以后面补充上。

File: examples/static/gpt/args.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+    parser.add_argument(
+        "--model_name_or_path",
+        default=None,
+        type=str,
+        required=True,
+        help="Path to pre-trained model or shortcut name selected in the list: ")
```
Comment (sandyhouse) [inline examples/static/gpt/args.py:40]: 同上。

File: examples/static/gpt/args.py [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+        "--use_sharding",
+        type=str2bool,
+        nargs='?',
+        const=False,
+        help="Use sharding Parallelism to training.")
+    parser.add_argument(
+        "--sharding_degree",
+        type=int,
+        default=1,
+        help="Sharding degree. Share the parameters to many cards.")
```
Comment (sandyhouse) [inline examples/static/gpt/args.py:172]: Share --> Shard

File: examples/static/gpt/README.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+```
+mkdir data
+mv train.data.json_ids.npz data
+```
+
+## 飞桨4D混合并行训练
+飞桨4D混合并行，使用sharding、模型并行、流水线并行和数据并行策略，使得训练千亿参数规模的模型成为可能。在本示例中，我们提供了基于飞桨最新混合并行策略的GPT预训练模型。运行下面脚本，即可进行模型预训练对应模型：
+```shell
+sh run.sh "0,1" "./output/dp" --micro_batch_size 4 --global_batch_size 8 --mp_degree 1 --dp_degree 2 --pp_degree 1 --debug true
+```
```
Comment (sandyhouse) [inline examples/static/gpt/README.md:36]: 后面可以补充下各个参数的含义。

File: examples/static/gpt/dataset.py [hunk 13]
```diff
@@ -0,0 +1,436 @@
...
+                worker_init_fn=worker_init,
+                collate_fn=Tuple(Stack(), Stack(), Stack(), Stack(), Stack()),
+                return_list=False)
+        return data_loader
+
+    # Note, data should be broardcast to all devices.
+    # for train, valid, test, the distinct data num is topo.data_info.size
+    # data_worldsize * data_inner_times -> topo.world.size
+    train_data_loader = build_dataset(0, "train", args.micro_batch_size *
+                                      args.max_steps * topo.data_info.size)
```
Comment (sandyhouse) [inline examples/static/gpt/dataset.py:323]: For pipeline, should we use args.micro_batch_size*args.max_steps*topo.data_info.size*num_micro_batches?

File: examples/static/gpt/dataset.py [hunk 14]
```diff
@@ -0,0 +1,436 @@
...
+    _filename += '_{}_indexmap'.format(name)
+    _filename += '_{}ns'.format(num_samples)
+    _filename += '_{}sl'.format(seq_length)
+    doc_idx_filename = _filename + '_doc_idx.npy'
+    sample_idx_filename = _filename + '_sample_idx.npy'
+    shuffle_idx_filename = _filename + '_shuffle_idx.npy'
+
+    # Build the indexed mapping if not exist.
+    # TODO @ZHUI support multi-mechine, worker_index % len(cudas()) == 0
+    if worker_index == 0:
```
Comment (sandyhouse) [inline examples/static/gpt/dataset.py:59]: 如果是多机训练，其它节点是不是就拿不到数据了。这个地方是否应该为worker_index % gpus_per_nodes == 0?

File: examples/static/gpt/global_setting.py [hunk 15]
```diff
@@ -0,0 +1,4 @@
+def init_global():
+    global _global_parallel_strategy
+    global _global_process_mesh
+    global ROOT_MESH
```
Comment (sandyhouse) [inline examples/static/gpt/global_setting.py:4]: The above two statements are unnecessary for hybrid-parallelism.

File: examples/static/gpt/modeling.py [hunk 16]
```diff
@@ -0,0 +1,863 @@
...
+    def _prepare_qkv(self, query, key, value, use_cache=False, cache=None):
+        r"""
+        Prapares linear projected queries, keys and values for usage of subsequnt
+        multiple parallel attention. If `cache` is not None, using cached results
+        to reduce redundant calculations.
+        """
+        q = self.q_proj(query)
+
+        if global_setting._global_parallel_stratergy == "mp":
+            auto.shard_tensor(
```
Comment (sandyhouse) [inline examples/static/gpt/modeling.py:177]: Why use shard_tensor api for auto parallelel?

File: examples/static/gpt/modeling.py [hunk 17]
```diff
@@ -0,0 +1,863 @@
...
+                self.k_proj.weight, global_setting._global_process_mesh, dim_mapping=[-1, 1])
+
+        v = self.v_proj(value)
+
+        if global_setting._global_parallel_stratergy == "mp":
+            auto.shard_tensor(
+                self.v_proj.weight, global_setting._global_process_mesh, dim_mapping=[-1, 0])
+        elif global_setting._global_parallel_stratergy == "dp_mp":
+            auto.shard_tensor(
```
Comment (sandyhouse) [inline examples/static/gpt/modeling.py:226]: why use shard_tensor?

File: examples/static/gpt/modeling.py [hunk 18]
```diff
@@ -0,0 +1,863 @@
...
+        if global_setting._global_parallel_stratergy == "mp":
+            auto.shard_tensor(
+                self.out_proj.weight,
+                global_setting._global_process_mesh,
+                dim_mapping=[0, -1])
+        elif global_setting._global_parallel_stratergy == "dp_mp":
+            auto.shard_tensor(
+                self.out_proj.weight,
+                global_setting._global_process_mesh,
+                dim_mapping=[1, -1])
```
Comment (sandyhouse) [inline examples/static/gpt/modeling.py:314]: why use shard_tensor?

File: examples/static/gpt/modeling.py [hunk 19]
```diff
@@ -0,0 +1,863 @@
...
+        if self.normalize_before:
+            tgt = self.norm2(tgt)
+
+        if global_setting._global_parallel_stratergy == "mp":
+            auto.shard_tensor(
+                self.linear1.weight, global_setting._global_process_mesh,
+                dim_mapping=[-1, 0])
+        elif global_setting._global_parallel_stratergy == "dp_mp":
+            auto.shard_tensor(
+                self.linear1.weight, global_setting._global_process_mesh,
```
Comment (sandyhouse) [inline examples/static/gpt/modeling.py:532]: why use shard_tensor?

File: examples/static/gpt/modeling.py [hunk 20]
```diff
@@ -0,0 +1,863 @@
...
+
+        if global_setting._global_parallel_stratergy == "mp":
+            auto.shard_tensor(
+                self.word_embeddings.weight,
+                global_setting._global_process_mesh,
+                dim_mapping=[0, -1])
+        elif global_setting._global_parallel_stratergy == "dp_mp":
+            auto.shard_tensor(
+                self.word_embeddings.weight,
+                global_setting._global_process_mesh,
```
Comment (sandyhouse) [inline examples/static/gpt/modeling.py:636]: why use shard_tensor?

File: examples/static/gpt/modeling_utils/data/iterator.py [hunk 21]
```diff
@@ -0,0 +1 @@
+# Iterator for NLP Dataset
```
Comment (sandyhouse) [inline examples/static/gpt/modeling_utils/data/iterator.py:1]: What does this file for?

File: examples/static/gpt/modeling_utils/transformers/generation_utils.py [hunk 22]
```diff
@@ -0,0 +1,940 @@
...
+                format(eos_token_id))
+
+        self.min_length = min_length
+        self.eos_token_id = eos_token_id
+
+    def __call__(self, input_ids, logits):
+        cur_len = input_ids.shape[-1]
+        if cur_len < self.min_length:
+            logits[:, self.eos_token_id] = -1e9
+        return logits
```
Comment (sandyhouse) [inline examples/static/gpt/modeling_utils/transformers/generation_utils.py:940]: 这个文件有用到吗？

File: examples/static/gpt/modeling_utils/transformers/gpt/modeling.py [hunk 23]
```diff
@@ -0,0 +1,996 @@
...
+            tgt = F.gelu(tgt, approximate=True)
+        # layers.Print(tgt, message="linear1")
+        tgt = self.dropout2(self.linear2(tgt))
+        # layers.Print(tgt, message="dropout2")
+        tgt = residual + tgt
+        # layers.Print(tgt, message="residual + tgt")
+
+        if not self.normalize_before:
+            tgt = self.norm2(tgt)
+        # layers.Print(tgt, message="norm2 after")
```
Comment (sandyhouse) [inline examples/static/gpt/modeling_utils/transformers/gpt/modeling.py:533]: Delete these commented statements。

File: examples/static/gpt/modeling_utils/transformers/gpt/modeling_auto.py [hunk 24]
```diff
@@ -0,0 +1,881 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (sandyhouse) [inline examples/static/gpt/modeling_utils/transformers/gpt/modeling_auto.py:1]: 删掉这个文件吧，手动混合并行里面不要混入自动并行的。

File: examples/static/gpt/modeling_utils/transformers/gpt/modeling_manual.py [hunk 25]
```diff
@@ -0,0 +1,919 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (sandyhouse) [inline examples/static/gpt/modeling_utils/transformers/gpt/modeling_manual.py:1]: modeling.py model_manual.py这两个文件什么区别?

File: examples/static/gpt/modeling_utils/transformers/gpt/modeling_serial.py [hunk 26]
```diff
@@ -0,0 +1,781 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (sandyhouse) [inline examples/static/gpt/modeling_utils/transformers/gpt/modeling_serial.py:1]: 该文件需要吗？

File: examples/static/gpt/modeling_utils/transformers/transformer/modeling.py [hunk 27]
```diff
@@ -0,0 +1,851 @@
...
+                start=0, end=src_max_len)
+
+        # Run encoder
+        src_emb = self.src_word_embedding(src_word)
+        src_pos_emb = self.src_pos_embedding(src_pos)
+        src_emb = src_emb + src_pos_emb
+        enc_input = F.dropout(
+            src_emb, p=self.dropout,
+            training=False) if self.dropout else src_emb
+        enc_output = self.transformer.encoder(enc_input, src_slf_attn_bias)
```
Comment (sandyhouse) [inline examples/static/gpt/modeling_utils/transformers/transformer/modeling.py:832]: 这个文件是必要的吗？

File: examples/static/gpt/modeling_utils/utils/downloader.py [hunk 28]
```diff
@@ -0,0 +1,338 @@
...
+    ('BERT-en-cased-large-whole-word-masking',
+     'https://bert-models.bj.bcebos.com/wwm_cased_L-24_H-1024_A-16.tar.gz'),
+    ('BERT-en-uncased-base',
+     'https://bert-models.bj.bcebos.com/uncased_L-12_H-768_A-12.tar.gz'),
+    ('BERT-en-uncased-large',
+     'https://bert-models.bj.bcebos.com/uncased_L-24_H-1024_A-16.tar.gz'),
+    ('BERT-en-cased-base',
+     'https://bert-models.bj.bcebos.com/cased_L-12_H-768_A-12.tar.gz'),
+    ('BERT-en-cased-large',
+     'https://bert-models.bj.bcebos.com/cased_L-24_H-1024_A-16.tar.gz'),
```
Comment (sandyhouse) [inline examples/static/gpt/modeling_utils/utils/downloader.py:93]: 这个文件是必要的吗？

File: examples/static/gpt/test_hybrid_parallel.py [hunk 29]
```diff
@@ -0,0 +1,404 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (sandyhouse) [inline examples/static/gpt/test_hybrid_parallel.py:1]: What does this file for？

**Review by sandyhouse — APPROVED** [review]
